Wrap insn_len for test mode and for CRuby

noahgibbs opened the pull request:
If we link against insn_len, a lot of other functions can be ported. If we link against insn_len, it breaks "cargo test". (Update: prevented it breaking cargo test, partly by not really testing it.)

Insn_len is also an interesting case, in that it needs a C wrapper (it's a static func) so that we can link against it. And it benefits from a Rust-side wrapper to prevent having to manually use "unsafe" everywhere. Would love opinions on whether this is the right way to wrap functions of this kind.

I've checked that this doesn't break "miniruby --yjit" any more than it already is, but it also doesn't call insn_len() yet.

noahgibbs committed:
Link against insn_len in prod after wrapping it from C; have it panic in test. Convert some funcs that need it.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -147,14 +162,12 @@ jit_obj_info_dump(codeblock_t *cb, x86opnd_t opnd) {
 {
-    return jit->insn_idx + insn_len(jit_get_opcode(jit));
+    (jit.get_insn_idx() + insn_len(jit.get_opcode())).try_into().unwrap()
 }
```
Here I think we should convert `insn_idx` to u32 for consistency because that's the type being used in `core.rs`.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -98,11 +98,26 @@ impl JITState {
+    pub fn set_record_boundary_patch_point(&mut self, val: bool)
+    {
+        self.record_boundary_patch_point = val;
+    }
```
I don't think there's much point in defining accessors for those because those fields are directly accessible within the `codegen.rs` module. Unless we decide to move `JITState` later 🤔

maximecb commented on yjit.c:
```diff
@@ -68,6 +68,13 @@ rb_yjit_alloc_exec_mem(uint32_t mem_size)
 }
 
+// Add a wrapper around this const static function so that Rust can see it.
+int
+rb_yarv_insn_len(VALUE insn)
+{
+    return insn_len(insn);
+}
```
Maybe we should have this return a uint32_t for convenience?

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -286,50 +299,42 @@ macro_rules! counted_exit {
-    mov(cb, scratch_reg, const_ptr_opnd(jit->pc + insn_len(jit->opcode)));
-    mov(cb, mem_opnd(64, REG_CFP, offsetof(rb_control_frame_t, pc)), scratch_reg);
+    let pc: *mut VALUE = jit.get_pc();
+    let ptr: *mut VALUE = unsafe { pc.offset(insn_len(jit.get_opcode()).try_into().unwrap()) };
```
Here I would break this expression down into parts so it's more clear what's happening. For example `let cur_insn_len = ...`.

noahgibbs replied:
Done.

maximecb commented on yjit/src/cruby.rs:
```diff
@@ -108,6 +108,13 @@ extern "C" {
+    unsafe {
+        raw_insn_len(VALUE(opcode)).try_into().unwrap()
+    }
+}
```
Not a bad idea to wrap it it but let's have it return a u32 for convenience.

maximecb commented:
Regarding the cargo test situation, we might have to just not test things that require linking, or always return 1 in test mode.

@XrXr would you have advice on this?

noahgibbs commented:
While we could shim insn_len in test mode to always return 1, shimming everything that crosses over from CRuby is going to get very restrictive very quickly. A *lot* is just not going to be testable. For starters, it means anything with ISEQs is back to either needing one single test-mode ISEQ (ew) or having us mock up ISEQs on the Rust side (duplicating a lot of CRuby structures or duplicating logic and using raw bytes.) Anything that reads or writes instance variables seems to use the ISEQ structures at least a bit, so that's going to be a *lot* of code.

Without CRuby callout, our tests will need to be very limited unit tests with all the real testing done using the linked CRuby binary. "Nearly all testing in Ruby" isn't so bad -- it's what we do in Ruby, obviously. But it would be nice if we could link CRuby somehow and test against it using Rust.

Might be worth seeing if we can build Ruby everything as static libs and link against that. It'd be a little weird since prod Ruby would need static Rust and test Rust would need static Ruby. But that sounds doable at first blush. I assume Rust can link against a static lib in test mode.

maximecb commented:
I agree shimming everything is going to be problematic.

I think building CRuby as a lib is impractical. I would vote for just not testing things that cross the CRuby boundary for now.

noahgibbs commented:
Okay. I've converted the u32s and removed the extra accessors. I'll figure out what we want to do with insn_len in test mode. Though at this point I don't think any of the tests call it, so "todo!()" might work.

noahgibbs commented:
I've re-removed the assert in jit_get_arg since it needs insn_len(). But with that gone, I can just define insn_len() as todo!() in test mode and everything passes.

maximecb commented:
I think you could define insn_len() as `panic!()` when in test mode.

maximecb commented:
You can also put a text string in there, `panic!("insn_len requires linking to CRuby, not available in test mode");`

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -147,14 +148,12 @@ jit_obj_info_dump(codeblock_t *cb, x86opnd_t opnd) {
+fn jit_next_insn_idx(jit: &JITState) -> u32
 {
-    return jit->insn_idx + insn_len(jit_get_opcode(jit));
+    (jit.insn_idx + insn_len(jit.get_opcode())).try_into().unwrap()
```
Hopefully we don't need `try_into().unwrap()` now that everything is u32? 🤞

noahgibbs replied:
I've remove a bunch of that. Let me know if you see anything else that should go.

maximecb commented on yjit/src/cruby.rs:
```diff
@@ -108,6 +108,20 @@ extern "C" {
     pub fn alloc_exec_mem(mem_size: u32) -> *mut u8;
 }
 
+#[cfg(not(test))]
```
You can also put the `#[cfg(not(test))]` directly in the `insn_len()` function, next to the expression it applies to.

noahgibbs replied:
True. Then I'd *also* probably want to turn off the "unreachable" warnings on the following code, or turn all the code on or off expression-by-expression. It winds up (to my eye) a lot messier than just having two single-expression functions.

maximecb replied:
I think I'd prefer wrapping the two blocks inside the function because it avoids duplicating the signature.

```
pub fn insn_len(opcode:usize) -> u32
{
    #[cfg(test)]
    panic!("insn_len is a CRuby function, and we don't link against CRuby for Rust testing!");

    #[cfg(not(test))]
    unsafe {
        raw_insn_len(VALUE(opcode))
    }
}
```

maximecb commented on yjit/src/cruby.rs:
```diff
@@ -99,7 +99,7 @@ extern "C" {
 
     #[link_name = "rb_yarv_insn_len"]
-    pub fn insn_len(v: VALUE) -> std::os::raw::c_int;
+    pub fn raw_insn_len(v: VALUE) -> u32;
```
I think you should leave the syntax as `c_int` and leave the C export as int, but do the casting in your wrapper:

`pub fn insn_len(opcode:usize) -> u32`

Sorry about the confusion. I suggested changing the C API earlier but I actually think it's best to change only the rust wrapper.

noahgibbs committed:
Make the jit_get_arg assertion non-test-only

noahgibbs commented:
Okay. Should be up to date w/ latest feedback. I've also rebased the commits into a cleaner history.

XrXr commented:
Re: linking for testing, I think it's doable, but not what we want at least for now. Doing so makes us test things outside of the normal lifecycle of Ruby, making it so that we don't necessary test the right thing. We could have special modes where we expose a Ruby method like `::YJIT.test_internals` and call that similar to `make btest` for cross cutting concerns.

maximecb commented:
Very nice :)

maximecb commented:
I like the simplicity of the current setup and that we can test the rust crate easily on its own. Hopefully we'll be running tests with `miniruby` next week.
